File: 98_Validate_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.validateTree(root, None, None) 
    
    def validateTree(self, root, low, high):
        if(root == None):
            return True 
        
        if((low != None and low >= root.val) or (high != None and high <= root.val)):
            return False 
        return self.validateTree(root.left, low, root.val) and self.validateTree(root.right, root.val, high)
        

File: 98_Validate_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return validateTree(root, -1, -1); 
    }
    bool validateTree(TreeNode* root, int left, int right)
    {
        if(root == NULL) return true;
        if((left != -1 && left >= root->val) || (right != -1 && right <= root->val))
        {
            return false; 
        }
        return validateTree(root->left, left, root->val) && validateTree(root->right, root->val, right); 
    }
};